apps/chat/tests/test_tools.py

```diff
@@ -0,0 +1,66 @@
+from datetime import datetime
+
+import pytest
+import pytz
+from freezegun import freeze_time
+
+from apps.chat.agent.schemas import WeekdaysEnum
+from apps.chat.agent.tools import UpdateScheduledMessageTool, _move_datetime_to_new_weekday_and_time
+from apps.events.models import ScheduledMessage
+from apps.utils.factories.events import EventActionFactory
+from apps.utils.factories.experiment import ExperimentSessionFactory
+from apps.utils.time import pretty_date
+
+
+@pytest.mark.parametrize(
+    ("initial_datetime_str", "new_weekday", "new_hour", "new_minute", "expected_new_datetime_str"),
+    [
+        ("2024-05-08 08:00:00", 2, 12, 0, "2024-05-08 12:00:00"),  # Only time change
+        ("2024-05-08 08:00:00", 0, 8, 0, "2024-05-06 08:00:00"),  # Wednesday to Monday
+        ("2024-05-29 08:00:00", 4, 8, 0, "2024-05-31 08:00:00"),  # Wednesday to Friday
+        ("2024-06-01 08:00:00", 0, 8, 0, "2024-05-27 08:00:00"),  # Saturday to Monday
+        ("2024-06-02 08:00:00", 0, 8, 0, "2024-05-27 08:00:00"),  # Sunday to Monday
+        ("2024-06-02 08:00:00", 1, 8, 0, "2024-05-28 08:00:00"),  # Sunday to Tuesday
+    ],
+)
+def test_move_datetime_to_new_weekday_and_time(
+    initial_datetime_str, new_weekday, new_hour, new_minute, expected_new_datetime_str
+):
+    """Test weekday and time changes. A weekday change will cause the datetime to jump to a different week"""
```

Comment:
I'm assuming behaviour here: Suppose it's Tuesday the second today and a scheduled messages is scheduled to go off on Friday but the user wants a monday schedule, then by changing it to mondays, the trigger date will shift to the Monday the 1st, effectively triggering the scheduled message. So basically: a preference update **should not delay a schedule** more than a week. If we allowed this, then a user can update their preferences in such a way that they'll never get the scheduled message.